src/Libraries/Microsoft.Extensions.AI/Functions/AIFunctionFactory.cs

```diff
@@ -432,42 +450,38 @@ static bool IsAsyncMethod(MethodInfo method)
+                    return async (taskObj, cancellationToken) =>
                     {
-                        var task = (Task)ReflectionInvoke(valueTaskAsTask, ThrowIfNullResult(result), null)!;
+                        var task = (Task)ReflectionInvoke(valueTaskAsTask, ThrowIfNullResult(taskObj), null)!;
                         await task.ConfigureAwait(false);
-                        return ReflectionInvoke(asTaskResultGetter, task, null);
+                        object? result = ReflectionInvoke(asTaskResultGetter, task, null);
+                        return await SerializeResultAsync(result, returnTypeInfo, cancellationToken).ConfigureAwait(false);
                     };
                 }
             }
 
-            // For everything else, just use the result as-is.
-            return result => new ValueTask<object?>(result);
-
-            // Throws an exception if a result is found to be null unexpectedly
-            static object ThrowIfNullResult(object? result) => result ?? throw new InvalidOperationException("Function returned null unexpectedly.");
-        }
+            // For everything else, just serialize the result as-is.
+            returnTypeInfo = serializerOptions.GetTypeInfo(returnType);
+            return (result, cancellationToken) => SerializeResultAsync(result, returnTypeInfo, cancellationToken);
 
-        /// <summary>Invokes the MethodInfo with the specified target object and arguments.</summary>
-        private static object? ReflectionInvoke(MethodInfo method, object? target, object?[]? arguments)
-        {
-#if NET
-            return method.Invoke(target, BindingFlags.DoNotWrapExceptions, binder: null, arguments, culture: null);
-#else
-            try
+            static async ValueTask<object?> SerializeResultAsync(object? result, JsonTypeInfo returnTypeInfo, CancellationToken cancellationToken)
             {
-                return method.Invoke(target, BindingFlags.Default, binder: null, arguments, culture: null);
-            }
-            catch (TargetInvocationException e) when (e.InnerException is not null)
-            {
-                // If we're targeting .NET Framework, such that BindingFlags.DoNotWrapExceptions
-                // is ignored, the original exception will be wrapped in a TargetInvocationException.
-                // Unwrap it and throw that original exception, maintaining its stack information.
-                System.Runtime.ExceptionServices.ExceptionDispatchInfo.Capture(e.InnerException).Throw();
-                return null;
+                if (returnTypeInfo.Kind is JsonTypeInfoKind.None)
+                {
+                    // Special-case trivial contracts to avoid the more expensive general-purpose serialization path.
+                    return JsonSerializer.SerializeToElement(result, returnTypeInfo);
+                }
+
+                // Serialize asynchronously to support potential IAsyncEnumerable responses.
```

Comment:
Should we add to STJ a SerializeToElementAsync?

Reply:
It's possible. Currently the only async methods we expose target `Stream` although we could conceivably expose variants for all other encoding mediums.